public/app/features/dashboard/containers/NewDashboardWithDS.tsx

```diff
@@ -0,0 +1,43 @@
+import React, { useEffect, useState } from 'react';
+
+import { getDataSourceSrv, locationService } from '@grafana/runtime';
+import { Page } from 'app/core/components/Page/Page';
+import { GrafanaRouteComponentProps } from 'app/core/navigation/types';
+
+import { getNewDashboardModelData, setDashboardToFetchFromLocalStorage } from '../state/initDashboard';
+
+export default function NewDashboardWithDS(props: GrafanaRouteComponentProps<{ datasourceUid: string }>) {
+  const [error, setError] = useState<string | null>(null);
+  const { datasourceUid } = props.match.params;
+
+  useEffect(() => {
+    const ds = getDataSourceSrv().getInstanceSettings(datasourceUid);
+    if (!ds) {
+      setError('Data source not found');
+      return;
+    }
+
+    const newDashboard = getNewDashboardModelData();
+    const { dashboard } = newDashboard;
+    dashboard.panels[0] = {
+      ...dashboard.panels[0],
+      datasource: {
+        uid: ds.uid,
+        type: ds.type,
+      },
+    };
+
+    setDashboardToFetchFromLocalStorage(newDashboard);
+    locationService.push('/dashboard/new');
```

Comment:
I think we should use `.replace()` here (otherwise the user cannot navigate back to `/datasources` from the datasources page.

Reply:
@leventebalogh ah good catch! will fix